File: optimizers/OBGWO.m
 function [Alpha_score,Alpha_pos]=OBGWO(SearchAgents_no,Max_iteration,lb,ub,dim,fobj,run);
 
     OP=0.1;
     
 %initialization of the population uniformly
  Positions=initialization(SearchAgents_no,dim,ub,lb,run);
  
  %evaluate their fitness
  for i=1:SearchAgents_no
        Fitness(i)=fobj(Positions(i,:));
  end
  fitness=Fitness';
  
  %GENERATE POP WITH OPPOSITION
  
  for i=1:SearchAgents_no
       for j=1:dim
            O_Positions(i,j)=(ub+lb)-Positions(i,j);
           
             if O_Positions(i,j)>ub 
                O_Positions(i,j)=ub;
             end
             if O_Positions(i,j)<lb
                O_Positions(i,j)=lb;
            
             end
       end
     O_Fitness(i)=fobj(O_Positions(i,:));
  end
  O_fitness=O_Fitness';
  
  %MERGE THESE TWO POP TO PICK INITIAL  POPULATION
  
        New_mat1=[fitness Positions;O_fitness O_Positions];
        [values, order]=sort(New_mat1(:,1));
        New_mat2=New_mat1(order,:);
        
        Positions=New_mat2(1:SearchAgents_no,2:size(New_mat2,2));
        fitness=New_mat2(1:SearchAgents_no,1);
        
        %SELECT THE LEADERS OF GWO
        
        Alpha_score= fitness(1,1); % Update alpha
        Alpha_pos= Positions(1,:);
            
        Beta_score= fitness(2,1); % Update alpha
        Beta_pos= Positions(2,:);
       
        Delta_score= fitness(3,1); % Update alpha
        Delta_pos= Positions(3,:);
        
        Saved_matrix=[Alpha_score Alpha_pos;Beta_score Beta_pos;Delta_score Delta_pos];
        
  l=0;
  
  while l<Max_iteration

   
         a=2-2*(l/Max_iteration);   % linearly decreasing vector from 2 to 0

    % Update the Position of search agents including omegas
    for i=1:size(Positions,1)
        for j=1:size(Positions,2)     
                       
            r1=rand(); 
            r2=rand(); 
            
            A1=2*a*r1-a; 
            C1=2*r2; 
            
            D_alpha=abs(C1*Alpha_pos(j)-Positions(i,j)); 
            X1=Alpha_pos(j)-A1*D_alpha; 
                       
            r1=rand();
            r2=rand();
            
            A2=2*a*r1-a; 
            C2=2*r2; 
            
            D_beta=abs(C2*Beta_pos(j)-Positions(i,j)); 
            X2=Beta_pos(j)-A2*D_beta;       
            
            r1=rand();
            r2=rand(); 
            
            A3=2*a*r1-a; 
            C3=2*r2; 
            
            D_delta=abs(C3*Delta_pos(j)-Positions(i,j)); 
            X3=Delta_pos(j)-A3*D_delta;             

            U_Positions(i,j)=(X1+X2+X3)/3;
   
             if U_Positions(i,j)>ub 
                U_Positions(i,j)=ub;
             end
             if U_Positions(i,j)<lb
                U_Positions(i,j)=lb;
            
             end
         end
          U_Fitness(i)=fobj(U_Positions(i,:));
    end
         U_fitness=U_Fitness';
         
         %UPDATED POSITIONS WITH GWO
         U_matrix=[U_fitness U_Positions];
         
  if rand()<=OP
      
    for i=1:SearchAgents_no
        for j=1:dim
             O_Positions(i,j)=(ub+lb)-Positions(i,j);
           
             if O_Positions(i,j)>ub 
                O_Positions(i,j)=ub;
             end
             if O_Positions(i,j)<lb
                O_Positions(i,j)=lb;
            
             end
       end
       O_Fitness(i)=fobj(O_Positions(i,:));
    end
    O_fitness=O_Fitness';
  
    %OPPOSITION MATRIX
    
       O_matrix=[O_fitness O_Positions];
       
        mat1=[U_matrix;O_matrix;Saved_matrix];
        [values, order]=sort(mat1(:,1));
        mat2=mat1(order,:);
        
      Alpha_score=mat2(1,1);
      Beta_score=mat2(2,1);
      Delta_score=mat2(3,1);
      Alpha_pos=mat2(1,2:size(New_mat2,2));
      Beta_pos=mat2(2,2:size(New_mat2,2));
      Delta_pos=mat2(3,2:size(New_mat2,2));
 
       
        New_mat1=[U_matrix;O_matrix];
        [values, order]=sort(New_mat1(:,1));
        New_mat2=New_mat1(order,:);
        
        Positions=New_mat2(1:SearchAgents_no,2:size(New_mat2,2));
        fitness=New_mat2(1:SearchAgents_no,1);
        
        %SAVED MATRIX
        Saved_matrix=[Alpha_score Alpha_pos;Beta_score Beta_pos;Delta_score Delta_pos];
  else
      
        mat1=[U_matrix;Saved_matrix];
        [values, order]=sort(mat1(:,1));
        mat2=mat1(order,:);
        
        %SELECT LEADERS
       Alpha_score=mat2(1,1);
      Beta_score=mat2(2,1);
      Delta_score=mat2(3,1);
      Alpha_pos=mat2(1,2:size(New_mat2,2));
      Beta_pos=mat2(2,2:size(New_mat2,2));
      Delta_pos=mat2(3,2:size(New_mat2,2));
      
      %SAVED_MATRIX
      Saved_matrix=[Alpha_score Alpha_pos;Beta_score Beta_pos;Delta_score Delta_pos];
      
      %POSITION FOR NEXT ITERATION
      Positions=U_Positions;
      fitness=U_fitness;
  end
  
%   
%   if rem(l,1)==0
%       display(num2str(Alpha_score));
%   end
  l=l+1;
  end
 end
        
        

  
   

File: Optimizers/Python/OBGWO.py
# matlab conversion from chatGPT

import numpy as np

class OBGWO:
    def __init__(self, fobj, lb, ub, dim, SearchAgents_no, Max_iteration, OP=0.1):
        self.fobj = fobj  # Objective function
        self.lb = lb  # Lower bounds
        self.ub = ub  # Upper bounds
        self.dim = dim  # Dimension of the problem
        self.SearchAgents_no = SearchAgents_no  # Number of search agents
        self.Max_iteration = Max_iteration  # Maximum number of iterations
        self.OP = OP  # Opposition probability
        
        # Initialize alpha, beta, delta
        self.Alpha_score = float("inf")
        self.Beta_score = float("inf")
        self.Delta_score = float("inf")
        
        self.Alpha_pos = np.zeros(dim)
        self.Beta_pos = np.zeros(dim)
        self.Delta_pos = np.zeros(dim)

    def initialization(self, SearchAgents_no, dim, ub, lb):
        return lb + (ub - lb) * np.random.rand(SearchAgents_no, dim)

    def optimize(self):
        # Initialization of positions
        Positions = self.initialization(self.SearchAgents_no, self.dim, self.ub, self.lb)
        
        # Evaluate initial fitness
        Fitness = np.array([self.fobj(ind) for ind in Positions])
        
        # Opposition-based learning
        O_Positions = self.ub + self.lb - Positions
        O_Positions = np.clip(O_Positions, self.lb, self.ub)
        O_Fitness = np.array([self.fobj(ind) for ind in O_Positions])
        
        # Merge populations and sort
        New_mat = np.vstack((np.hstack((Fitness.reshape(-1, 1), Positions)),
                             np.hstack((O_Fitness.reshape(-1, 1), O_Positions))))
        New_mat = New_mat[np.argsort(New_mat[:, 0])]
        
        # Update positions and fitness
        Positions = New_mat[:self.SearchAgents_no, 1:]
        Fitness = New_mat[:self.SearchAgents_no, 0]
        
        # Initialize Alpha, Beta, Delta
        self.update_leaders(Fitness, Positions)
        Saved_matrix = np.vstack((self.Alpha_score, self.Alpha_pos,
                                  self.Beta_score, self.Beta_pos,
                                  self.Delta_score, self.Delta_pos)).reshape(3, -1)
        
        l = 0
        while l < self.Max_iteration:
            a = 2 - 2 * l / self.Max_iteration  # Decrease a linearly
            
            # Update positions of search agents
            U_Positions = self.update_positions(Positions, a)
            U_Positions = np.clip(U_Positions, self.lb, self.ub)
            U_Fitness = np.array([self.fobj(ind) for ind in U_Positions])
            
            # Create updated matrix
            U_matrix = np.hstack((U_Fitness.reshape(-1, 1), U_Positions))
            
            if np.random.rand() <= self.OP:
                # Opposition-based learning
                O_Positions = self.ub + self.lb - Positions
                O_Positions = np.clip(O_Positions, self.lb, self.ub)
                O_Fitness = np.array([self.fobj(ind) for ind in O_Positions])
                
                # Merge and sort the matrices
                O_matrix = np.hstack((O_Fitness.reshape(-1, 1), O_Positions))
                mat = np.vstack((U_matrix, O_matrix, Saved_matrix))
                mat = mat[np.argsort(mat[:, 0])]
                
                # Update leaders
                self.update_leaders(mat[:, 0], mat[:, 1:])
                
                # Update saved matrix
                Saved_matrix = mat[:3, :]
                
                # Select new positions for next iteration
                Positions = mat[:self.SearchAgents_no, 1:]
                Fitness = mat[:self.SearchAgents_no, 0]
            else:
                # Merge U_matrix with saved matrix
                mat = np.vstack((U_matrix, Saved_matrix))
                mat = mat[np.argsort(mat[:, 0])]
                
                # Update leaders
                self.update_leaders(mat[:, 0], mat[:, 1:])
                
                # Update saved matrix
                Saved_matrix = mat[:3, :]
                
                # Positions for next iteration
                Positions = U_Positions
                Fitness = U_Fitness
            
            l += 1
            
        return self.Alpha_score, self.Alpha_pos
    
    def update_leaders(self, fitness, positions):
        self.Alpha_score = fitness[0]
        self.Alpha_pos = positions[0, :]
        
        self.Beta_score = fitness[1]
        self.Beta_pos = positions[1, :]
        
        self.Delta_score = fitness[2]
        self.Delta_pos = positions[2, :]
    
    def update_positions(self, Positions, a):
        r1, r2 = np.random.rand(), np.random.rand()
        A1, C1 = 2 * a * r1 - a, 2 * r2
        X1 = self.Alpha_pos - A1 * np.abs(C1 * self.Alpha_pos - Positions)
        
        r1, r2 = np.random.rand(), np.random.rand()
        A2, C2 = 2 * a * r1 - a, 2 * r2
        X2 = self.Beta_pos - A2 * np.abs(C2 * self.Beta_pos - Positions)
        
        r1, r2 = np.random.rand(), np.random.rand()
        A3, C3 = 2 * a * r1 - a, 2 * r2
        X3 = self.Delta_pos - A3 * np.abs(C3 * self.Delta_pos - Positions)
        
        return (X1 + X2 + X3) / 3


